src/geom/prep/PreparedPolygonPredicate.cpp

```diff
@@ -37,71 +39,116 @@ namespace prep { // geos.geom.prep
 //
 // protected:
 //
-bool
-PreparedPolygonPredicate::isAllTestComponentsInTarget(const geom::Geometry* testGeom) const
-{
-    geom::Coordinate::ConstVect pts;
-    geom::util::ComponentCoordinateExtracter::getCoordinates(*testGeom, pts);
-
-    for(std::size_t i = 0, ni = pts.size(); i < ni; i++) {
-        const geom::Coordinate* pt = pts[i];
-        const int loc = prepPoly->getPointLocator()->locate(pt);
-        if(geom::Location::EXTERIOR == loc) {
-            return false;
+struct AnyMatchingLocationFilter : public GeometryComponentFilter {
+    explicit AnyMatchingLocationFilter(algorithm::locate::PointOnGeometryLocator* locator, int loc) :
+        pt_locator(locator), test_loc(loc), found(false) {}
+
+    algorithm::locate::PointOnGeometryLocator* pt_locator;
+    const int test_loc;
+    bool found;
+
+    void filter_ro(const Geometry* g) override {
+        const Coordinate* pt = g->getCoordinate();
+        const int loc = pt_locator->locate(pt);
+
+        if (loc == test_loc) {
+            found = true;
+        }
+    }
+
+    bool isDone() override {
+        return found;
+    }
+};
+
+struct AnyNotMatchingLocationFilter : public GeometryComponentFilter {
+    explicit AnyNotMatchingLocationFilter(algorithm::locate::PointOnGeometryLocator* locator, int loc) :
+            pt_locator(locator), test_loc(loc), found(false) {}
+
+    algorithm::locate::PointOnGeometryLocator* pt_locator;
+    const int test_loc;
+    bool found;
+
+    void filter_ro(const Geometry* g) override {
+        const Coordinate* pt = g->getCoordinate();
+        const int loc = pt_locator->locate(pt);
+
+        if (loc != test_loc) {
+            found = true;
+        }
+    }
+
+    bool isDone() override {
+        return found;
+    }
+};
+
+struct OutermostLocationFilter : public GeometryComponentFilter {
+    explicit OutermostLocationFilter(algorithm::locate::PointOnGeometryLocator* locator) :
+    pt_locator(locator),
+    max_loc(geom::Location::UNDEF),
+    found(false) {}
+
+    algorithm::locate::PointOnGeometryLocator* pt_locator;
+    Location::Value max_loc;
+    bool found;
+
+    void filter_ro(const Geometry* g) override {
+        const Coordinate* pt = g->getCoordinate();
+        auto loc = static_cast<Location::Value>(pt_locator->locate(pt));
+
+        max_loc = std::max(max_loc, loc);
```

Comment:
Not sure it's good to rely on numeric ordering for locations.

Reply:
Yes, I had some hesitation on this. I changed it just now in fa665c2.

A cleaner implementation might be to continue w/`stx::max` but add a compile-time `static_assert` to verify that the locations are ordered in the expected way. I don't think that would be portable to Java, though.

Reply:
Maybe good enough to leave as is, and put a note in Location that values should be ordered.

Even better might be a comparison method on Location which highlights that the ordering is expected.

Reply:
Since fa665c2 it no longer depends on ordering.